topdown/cache/cache.go

```diff
@@ -165,3 +170,14 @@ func (c *cache) maxSizeBytes() int64 {
 	}
 	return *c.config.InterQueryBuiltinCache.MaxSizeBytes
 }
+
+type keyMutex struct {
+	sync.Map
+}
+
+func (m *keyMutex) Lock(k ast.Value) func() {
+	value, _ := m.LoadOrStore(k, &sync.Mutex{})
```

Comment:
A `sync.Mutex` probably isn't large, but this means that for each `k` ever encountered, we'll store one. And we'll never drop it. So if some long-running OPA instances uses, say
```rego
allow if http.send({"url": "https://foo", "method": "POST", "headers": {"now": time.now() }}}.body.allowed
```

we would keep accumulating these. Is that a problem? 💭

Reply:
It means that any OPA instance that does unique requests will _eventually_ run out of memory. You also need to consider the size of the cache key itself. 

In my company's case, we're using oauth bearer tokens in our requests from OPA. These are just below 1kb, and are renewed every now and then

Reply:
Given the cache key size the impact of not dropping keys from the cache should be low I imagine but there could certainly be a case where a long running OPA could eventually OOM due to it. 

The original change was:

```go
for key := c.l.Front(); key != nil && (c.usage+size > limit); key = c.l.Front() {
    dropKey := key.Value.(ast.Value)
    c.unsafeDelete(dropKey)
    c.l.Remove(key)
    c.keyMtx.Delete(dropKey)                     //  <--------this is change
    dropped++
}
```

@asleire raised the question: _Can this to remove a mutex that is currently locked by an in-progress request, allowing a new concurrent request with the same key to take place?_

So let's say there are 2 concurrent requests for the key `foo`. Both routines grab the value (ie. `sync.Mutex`) from the `sync.Map` and one of them takes the lock. Then this routine say while inserting a value for key `foo` in the cache ends up removing an existing stale value for `foo` from the cache and also removes  key `foo` from the `sync.Map`. But the other routine already is blocked on the mutex so it should get unblocked once the 1st go routine finishes its work. In this case deleting key `foo` from the `sync.Map` should be ok.
 The edge case here is if the 1st routine deletes  key `foo` from the `sync.Map` before the 2nd could read it and if that happens we've got the original issue again.

One solution @asleire suggested is to reduce the cache size as in [this](https://github.com/open-policy-agent/opa/pull/5361/files#diff-b8d8349b8c6177aedd00ad14696a2225d2b275061b62dcfb39e6a41f652149d7R141) change.


Another solution would be not to remove the current key from the `sync.Map`. Something like:

```go
func (c *cache) unsafeInsert(k ast.Value, v InterQueryCacheValue) (dropped int) {
	size := v.SizeInBytes()
	limit := c.maxSizeBytes()

	if limit > 0 {
		if size > limit {
			dropped++
			return dropped
		}

		for key := c.l.Front(); key != nil && (c.usage+size > limit); key = c.l.Front() {
			dropKey := key.Value.(ast.Value)
			c.unsafeDelete(dropKey)
			c.l.Remove(key)

			if dropKey.Compare(k) != 0 {
				c.keyMtx.Delete(dropKey)
			}

			dropped++
		}
	}

	c.items[k.String()] = v
	c.l.PushBack(k)
	c.usage += size
	return dropped
}
```

The edge case here would be a concurrent request for `dropKey`. 

Any thoughts?

Reply:
I'm thinking of the edge case where 3 request occur near-simultaneously:
1 request with key `foo`
2 requests with key `bar`

In this scenario the cache is already full.
1. `foo` request is completed, and the request is about to insert into cache
2. the first `bar` request is about to start, the request gets a lock on `keyMtx(foo)`
3. `foo` request has to delete entries from the cache, it removes the mutex `keyMtx(foo)`
4. the second `bar` request is about to start, since `keyMtx` doesnt contain a mutex for `foo` it creates a new mutex and locks it
5. both `bar` requests are now running concurrently and upon cache insertion they will both increase the `usage` counter, we have the same problem as we're trying to fix